src/app/outbound.rs

```diff
@@ -157,85 +157,111 @@ impl tap::Inspect for Endpoint {
 }
 
 pub mod discovery {
-    use futures::{future::Future, Async, Poll};
-    use std::net::SocketAddr;
-
     use super::super::dst::DstAddr;
     use super::Endpoint;
-    use control::destination::Metadata;
+    use control::destination::{Metadata, Unresolvable};
+    use futures::{future::Future, Async, Poll};
     use proxy::{http::settings, resolve};
     use transport::tls;
     use {Addr, Conditional, NameAddr};
 
-    #[derive(Clone, Debug)]
-    pub struct Resolve<R: resolve::Resolve<NameAddr>>(R);
+    #[derive(Clone)]
+    pub struct Resolver<R: resolve::Resolve<NameAddr>>(R);
```

Comment:
I feel like it _might_ be worth considering just moving all the remaining functionality in this module into `control::destination::Resolver`. It seems like this module is just a pretty thin layer over the `destination` resolver and we could probably put all that functionality in one place. I think a majority of the code here now is just re-implementing some future combinators, and there's only a few lines of actual _logic_ here.

On the other hand, moving that to `control::destination` might be a worse separation of concerns. 🤷‍♀